pool/context_pool.go

```diff
@@ -23,14 +25,31 @@ type ContextPool struct {
 // are busy, a call to Go() will block until the task can be started.
 func (p *ContextPool) Go(f func(ctx context.Context) error) {
 	p.errorPool.Go(func() error {
-		err := f(p.ctx)
-		if err != nil && p.cancelOnError {
+		// If we aren't cancelling on error, just return the result of f.
```

Comment:
Alternatively, why not just add a defer to the top of this closure?

```go
func (p *ContextPool) Go(f func(ctx context.Context) error) {
	p.errorPool.Go(func() error {
		defer func() {
			if p.cancelOnError {
				if r := recover(); r != nil {
					p.cancel()
					panic(r)
				}
			}
		}()

		err := f(p.ctx)
		if err != nil && p.cancelOnError {
			// Leaky abstraction warning: We add the error directly because
			// otherwise, canceling could cause another goroutine to exit and
			// return an error before this error was added, which breaks the
			// expectations of WithFirstError().
			p.errorPool.addErr(err)
			p.cancel()
			return nil
		}
		return err
	})
}
```

Reply:
Doh! of course - updated in 2a37a2a2d3c840b9100d8cfcfbd0da33fc1cc078